### PR TITLE
11 req parsing handle chunked encoding

### LINKED ISSUE #11: req_parsing: handle chunked encoding
The body of HTTP request can be encoded in chunks according to RFC 7230. This scenario needs to be handled by the RequestParser.

### PR DESCRIPTION
**Pull request by u413-284-si**
Introduces ability to parse chunked bodies:

Chunked body:
- indicated by "Transfer-Encoding: chunked"
- format:
- chunk size in hex representation and CRLF
- chunk data and CRLF
- read and convert chunk size
- read chunk data and check if chunk size corresponds to actual chunk size
- end of chunked body indicated by zero-size chunk = break condition of loop

Non-chunked body:
- indicated by "Content-Length" (can be overwritten by Transfer-Encoding, then revert to chunked body)
- fix parsing to insert newlines at correct positions while reading the body line by line

Generally fix interpretation of header fields "Transfer-Encoding" and "Content-Length" according to RFC 7230 and RFC 9112.

Add tests for body parsing.

TLDR: Enhance HTTP request parser with ability to read chunked and non-chunked bodies.

close #11

### CODE REVIEW

File: src/RequestParser.cpp [hunk 1]
```diff
@@ -163,9 +163,34 @@ bool	RequestParser::isValidHeaderFieldNameChar(uint8_t c) const
 	}
 }
 
+size_t	RequestParser::convertHex(const std::string& chunkSize) const
```
Comment (gwolf-011235) [inline src/RequestParser.cpp:166]: Missing Doxygen

File: src/RequestParser.cpp [hunk 2]
```diff
@@ -445,6 +469,69 @@ std::string	RequestParser::parseVersion(const std::string& requestLine)
...
+		length += numChunkSize;
+		std::getline(requestStream, strChunkSize);
+		if (strChunkSize[strChunkSize.size() - 1] == '\r')
+			strChunkSize.erase(strChunkSize.size() - 1);
+		else {
+			m_errorCode = 400;
+			throw std::runtime_error(ERR_MISS_CRLF);
+		}
+		numChunkSize = convertHex(strChunkSize);
+	}
```
Comment (gwolf-011235) [inline src/RequestParser.cpp:508]: I think this could be reduced using 
``` 
do { 
  ... 
} while ( numChunkSize > 0 )
```

File: src/RequestParser.cpp [hunk 3]
```diff
@@ -445,6 +469,69 @@ std::string	RequestParser::parseVersion(const std::string& requestLine)
...
+}
+
+void	RequestParser::parseNonChunkedBody(std::istringstream& requestStream)
```
Comment (gwolf-011235) [inline src/RequestParser.cpp:511]: Missing Doxygen

**Review by gwolf-011235 — CHANGES_REQUESTED** [review]
Looks good!

Mostly NITs but one important thing: we don't use length in parseChunkedBody(). I think it's because we would need to check with config_file? I would add a FIXME comment

File: src/RequestParser.cpp [hunk 4]
```diff
@@ -445,6 +469,69 @@ std::string	RequestParser::parseVersion(const std::string& requestLine)
 	return (requestLine.substr(++i));
 }
 
+
+void	RequestParser::parseChunkedBody(std::istringstream& requestStream)
+{
+	int			length = 0;
+	std::string	strChunkSize;
+	std::getline(requestStream, strChunkSize);
+	if (strChunkSize[strChunkSize.size() - 1] == '\r')
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/RequestParser.cpp:478]: It's only available in C++20

File: src/RequestParser.cpp [hunk 5]
```diff
@@ -445,6 +469,69 @@ std::string	RequestParser::parseVersion(const std::string& requestLine)
 	return (requestLine.substr(++i));
 }
 
+
+void	RequestParser::parseChunkedBody(std::istringstream& requestStream)
+{
+	int			length = 0;
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/RequestParser.cpp:475]: It may be used to replace a value given by the header "Content-Length":

If a message is received with both a Transfer-Encoding and a Content-Length header field, the
Transfer-Encoding overrides the Content-Length. (RFC 9112, 6.3, p.19)

Question is: do we still do anything with that value afterwards in the response builder or elsewhere or can it be discarded?

File: src/RequestParser.cpp [hunk 6]
```diff
@@ -445,6 +469,69 @@ std::string	RequestParser::parseVersion(const std::string& requestLine)
 	return (requestLine.substr(++i));
 }
 
+
+void	RequestParser::parseChunkedBody(std::istringstream& requestStream)
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/RequestParser.cpp:473]: Added.

File: src/RequestParser.cpp [hunk 7]
```diff
@@ -445,6 +469,69 @@ std::string	RequestParser::parseVersion(const std::string& requestLine)
...
+	std::string body;
+	size_t		length = 0;
+
+	while (std::getline(requestStream, body)) {
+		length += body.size() + 1;
+		if (body[body.size() - 1] == '\r')
+			body.erase(body.size() - 1);
+		if (!m_request.body.empty())
+			m_request.body += '\n';
+		m_request.body += body;
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/RequestParser.cpp]: I have fixed it to differentiate whether the last line was read by getline with a newline (in this case add newline) or with EOF (in this case omit newline).

File: src/RequestParser.cpp [hunk 8]
```diff
@@ -445,6 +469,69 @@ std::string	RequestParser::parseVersion(const std::string& requestLine)
...
+	}
+	size_t	contentLength = 0;
+	std::istringstream	iss(m_request.headers["Content-Length"]);
+	iss >> contentLength;
+	if (iss.fail())
+		throw std::runtime_error(ERR_CONVERSION_STRING_TO_SIZE_T);
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/RequestParser.cpp:575]: True, changed.

File: src/RequestParser.cpp [hunk 9]
```diff
@@ -538,14 +625,18 @@ void	RequestParser::checkTransferEncoding()
...
 			m_errorCode = 400;
-			throw std::runtime_error(ERR_NON_EXISTENT_CHUNKED_ENCODING);
+			throw std::runtime_error(ERR_NON_EXISTENT_TRANSFER_ENCODING);
 		}
 
-		std::vector<std::string>	encodings = split(m_request.headers["Transfer-Encoding"], ',');
-		if (encodings[encodings.size() - 1] != "chunked") {
-			m_errorCode = 400;
-			throw std::runtime_error(ERR_NON_FINAL_CHUNKED_ENCODING);
+		if (m_request.headers["Transfer-Encoding"].find("chunked") != std::string::npos) {
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/RequestParser.cpp:678]: All these checks happen only if the former check:
```
if (m_request.headers.find("Transfer-Encoding") != m_request.headers.end())
```
is true. So we can expect the element access to find the key.

File: src/RequestParser.cpp [hunk 10]
```diff
@@ -538,14 +625,18 @@ void	RequestParser::checkTransferEncoding()
...
+			std::vector<std::string>	encodings = split(m_request.headers["Transfer-Encoding"], ',');
+			if (encodings[encodings.size() - 1] != "chunked") {
```
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/RequestParser.cpp:680]: It's only available in C++20

File: src/RequestParser.cpp (hunk 1 above)
Comment (u413-284-si) [reply src/RequestParser.cpp:166]: Added

File: src/RequestParser.cpp (hunk 3 above)
Comment (u413-284-si) [reply src/RequestParser.cpp:556]: Added

File: src/RequestParser.cpp (hunk 2 above)
Comment (u413-284-si) [reply src/RequestParser.cpp]: True, changed.

**Review by QCHR1581 — APPROVED** [review]
Seems very well thought out to me. Tests work and changes were also made accordingly to the feedback of Gabriel.

File: src/RequestParser.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (gwolf-011235) [reply src/RequestParser.cpp:508]: It seems like we could decide to error? 
Right after it says "Such a message might indicate an attempt to perform request smuggling (Section 11.2) or response splitting (Section 11.1) and ought to be handled as an error. "
Although it's not a MUST. An intermediary MUST scrub the length if he wants to forward the message.

A thought re program logic: the server config decides what the client max body size is. I think we would need to check beforehand if we reached this limit already (else a malicious client could bombard us with data). But the head does not count towards the body size.
So I think we would need to parse the header, see that it is chunked, leave and come back for the next batch?

File: src/RequestParser.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (gwolf-011235) [reply src/RequestParser.cpp]: D`oh!

Well to my defense it has been available since C++11, and is constexpr since C++20 :D

If I understand correctly it was added to the 98 standard via DefectReport but not adopted by GNU libstdc++.
Compare with std::map::at() which was also later added and did get adopted. 

Can be resolved after enough bemusement did occur!

File: src/RequestParser.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (gwolf-011235) [reply src/RequestParser.cpp:681]: Yes.

I would say this falls into the constify everything approach of C++

If you can choose between mutating and non-mutating element access, why choose the mutating one?

File: src/RequestParser.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/RequestParser.cpp:508]: But they don´t specifiy what kind of error should be chosen. Probably 400, bad request again. 

What do you mean with the last sentence? 

The config might define a client request body size. Then we could check against this value with "length". But the config might not define it. Then we  would not need to count the length of the chunked body.  Also because we are not an intermediary which forwards the actual body length to an origin server.

File: src/RequestParser.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/RequestParser.cpp]: Where does it say about the Defect Report?

File: src/RequestParser.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/RequestParser.cpp]: We can also use .at() to make it even more robust. Changed.

File: src/RequestParser.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (gwolf-011235) [reply src/RequestParser.cpp:508]: If the config does not define it, we should set a default. For nginx it's 1m: [Link](https://nginx.org/en/docs/http/ngx_http_core_module.html#client_max_body_size)

The problem is in the current implementation we read however long from the request stream. This indicates that we already received everything from the client and are now stitching together the body. This means a client could send us unlimited amount of data and we would happily take it. Then we would try in here to stitch it together - if we not already crashed, because no memory left.
What I'm trying to say is:
- we need to know that the client message is ready to be parsed, which we know by content-length or chunked encoding
- but we also need a way to come back here if we did not receive everything already (in case of chunked encoding)

File: src/RequestParser.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (gwolf-011235) [reply src/RequestParser.cpp]: https://en.cppreference.com/w/cpp/string/basic_string/back in the last section Defect Reports:

_[...] The following behavior-changing defect reports were applied retroactively to previously published C++ 
[...] C++98: std::basic_string did not have the member function back() | added_

File: src/RequestParser.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/RequestParser.cpp:508]: The way I see it we stay in this loop as long as not all the data (chunks) have been transmitted (indicated by reception of  the zero-chunk) or we realise that the length of the received data exceeds the given max size of the config file (or the default of 1m - I guess that's one megabyte). For that we would need to implement a check in the loop to verify this hard boundery has not been hit  yet. I would use the already exisiting issue #11 to track this. The issue itself would be rewritten.

Don´t see why we would leave the function though...

File: src/RequestParser.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/RequestParser.cpp]: Oh I see!

File: src/RequestParser.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (gwolf-011235) [reply src/RequestParser.cpp:508]: Maybe I'm missing something.

RequestParser::parseHttpRequest receives a const std::string& request. It then checks if request line, headers and optional body are present (in this fixed order). 
This means RequestParser expects to receive the **full** request of the client. Therefore we have to receive and save as long as the client sends data. When the client doesn't send anymore only then can we go into the parser and check if this request is valid - because only then do we have the full request string.

I think we need to be able to partially check the request. The header gets parsed and the (first part of the) body. If our buffer can hold the full request we can parse it right away. If our buffer couldn't hold everything we need to validate the header and come back later when we received more. Or kick the client if he does stupid.

The server doesn't know if the request is valid since he can't parse the header. So he needs to accept everything

[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/RequestParser.cpp:508]: I see now what you mean. 
The current implementation demands that the RequestParser object indeed receives the full request string, also when the body is chunked it would expect to receive the entire chunked body including the terminating zero-chunk. The aggregation of the chunked pieces happens beforehand in the handleConnections() function. 

I have added a check on whether the http request is complete - 
firstly checking for complete header section, then looking for whether a body is present differentiating between chunked and non-chunked bodies. Only when the body is completely transmitted, the parser is then called and receives the full string. Otherwise the handleConnections() function exits and is called again when data is available and adds the new data to the request string.

In order to avoid an endless acceptance of data (faulty requests w/o clear indicators of the body end) we will also check against the max. body size as discussed previously. This feature shall be introduced after the configParser is implemented.
